code/ui/components/src/components/Button/Button.tsx

```diff
@@ -1,29 +1,121 @@
-import type { FC, ComponentProps, ReactNode } from 'react';
-import React, { forwardRef } from 'react';
+import type { ButtonHTMLAttributes, SyntheticEvent } from 'react';
+import React, { forwardRef, useEffect, useState } from 'react';
 import { styled } from '@storybook/theming';
 import { darken, lighten, rgba, transparentize } from 'polished';
+import { Slot } from '@radix-ui/react-slot';
+
+export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
```

Comment:
`ButtonHTMLAttributes<HTMLButtonElement>` isn't strictly correct is it, as `asChild` could turn this into any element? @kasperpeulen 

I don't know if there are types here to inspire us.
https://github.com/radix-ui/primitives/blob/main/packages/react/primitive/src/Primitive.tsx

Reply:
Now that I've slept on it this might not be super important, given that props to the child should just be set on the child itself and not on the parent button. so while this is not technically true, it's probably okay.

Reply:
It is inspired by shadcn: https://ui.shadcn.com/docs/components/button

```tsx
export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  asChild?: boolean
}
 
const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : "button"
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    )
  }
)
Button.displayName = "Button"
```

I think it is okay indeed.